rviz_mesh_tools_plugins/src/MeshDisplay.cpp

```diff
@@ -866,28 +871,58 @@ void MeshDisplay::processMessage(
   }
   setVertexNormals(normals);
 
-  // TODO
+  // TODO service stuff
   // std::cout << "requestVertexColors" << std::endl;
   // requestVertexColors(meshMsg.uuid);
   // std::cout << "requestMaterials" << std::endl;
   // requestMaterials(meshMsg.uuid);
   // std::cout << "done." << std::endl;
 }
 
-void MeshDisplay::incomingGeometry(
+void MeshDisplay::meshGeometryCallback(
   const mesh_msgs::msg::MeshGeometryStamped::ConstSharedPtr& meshMsg)
 {
   m_messagesReceived++;
   setStatus(rviz_common::properties::StatusProperty::Ok, "Topic", QString::number(m_messagesReceived) + " messages received");
   processMessage(*meshMsg);
+
+  // check for cached color and cost msgs that might have arrived earlier:
+  if (m_colorsMsgCache && m_colorsMsgCache->getOldestTime().seconds() != 0.0) {
+    // TODO: The check with getOldestTime() is a workaround to avoid segfault from bug in getSurroundingInterval() when cache is empty. 
+    // Remove when the changes from https://github.com/ros2/message_filters/pull/116 are released.
+    const auto colorMsgs = m_colorsMsgCache->getSurroundingInterval(meshMsg->header.stamp, meshMsg->header.stamp);
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("rviz_mesh_tools_plugins"), "Got " << colorMsgs.size() << " color msgs from cache");
+    for (const mesh_msgs::msg::MeshVertexColorsStamped::ConstSharedPtr& colorMsg : colorMsgs) {
+      if (colorMsg->uuid == meshMsg->uuid) {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("rviz_mesh_tools_plugins"), "Found cached color msg with matching UUID, applying its information now");
+        vertexColorsCallback(colorMsg);
+      } else {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("rviz_mesh_tools_plugins"), "Found cached color msg, but UUIDs do not match. Ignoring color msg.");
+      }
+    }
+  }
+  if (m_costsMsgCache && m_costsMsgCache->getOldestTime().seconds() != 0.0) {
+    // TODO: The check with getOldestTime() is a workaround to avoid segfault from bug in getSurroundingInterval() when cache is empty. 
+    // Remove when the changes from https://github.com/ros2/message_filters/pull/116 are released.
```

Comment:
This workaround is suboptimal, but the best I came up with. In simulation, when using sim time, it can cause colors or costs to not get displayed because its timestamp can actually be 0.0s.
In a non-simulation environment, the timestamp will never be 0.

While this sounds very unlikely at first, this regularly occurs in my setup where I launch both gazebo and rviz in one launch file, with `use_sim_time` enabled. Meh!

Reply:
I'm fine with that